### ISSUE TITLE
ValueTask.AsTask & ValueTask<TResult>.AsTask – useless remarks

### ISSUE THREAD

**Issue by Uzivatel919** [dotnet/dotnet-api-docs#7233]
Regard [ValueTask<TResult> Struct|Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=net-5.0#remarks) and [ValueTask Struct|Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask?view=net-5.0#remarks) is not recommended to call `AsTask` multiply times.

I made brief check with `Value`/`ValueTask<TResult>` and it seems this property belongs only to `Value`/`ValueTask<TResult>` of  `IValueTaskSource`/`IValueTaskSource<TResult>` tasks.

This fact should be stated in _Remarks_ instead of repeating what is stated in _Returns_ section. I would do it myself but I do not know details.

**Comment by svick** [comment]
As I understand it, this is because the framework wants to be free to change a `ValueTask` to use `IValueTaskSource` in a later version. If the documentation was changed the way you suggest, that would become a breaking change, which is why it should not change.

**Comment by Uzivatel919** [comment]
@svick, apologize my missing knowledge. I see [`ValueTask(IValueTaskSource, Int16)`](ValueTask(IValueTaskSource, Int16)). It is just dummy? What can be subject of change? Also general page denotes should need.

**Comment by svick** [comment]
@Uzivatel919 Imagine that .Net 6 had a method `ValueTask SomeMethod()`, which returns a `Task`-backed `ValueTask`. If I understand correctly what you're suggesting, then the documentation would claim that it's fine call `AsTask` multiple times on the returned `ValueTask`. (Which is technically correct.)

But then, in .Net 7, the implementation of `SomeMethod` changes, and now it returns an `IValueTaskSource`-backed `ValueTask`. The problem is, anyone who followed the documentation and called `AsTask` multiple times now has broken code.

That's why the documentation doesn't (and shouldn't) explain this detail: because people shouldn't be encouraged to write code that might become broken in the future.

**Comment by Uzivatel919** [comment]
@svick, I see. You understand me almost fully. What I suggest is not to denote ability to be fine with calling `AsTask` multiply times in path when it is sure it was created from _result_ or `Task`. I suggest to explain `IValueTaskSource` case as source of pattern-recommendation explaining or referring to information what is problem technically. I mean it is `IValueTaskSource` object reuse.

**Comment by scalablecory** [comment]
Callers generally don't know how their `ValueTask` was made, though this might be of some use to `internal` code.

We should be careful here because this is already a challenging API for people to use and presenting another "pivot point" for people to think about might be harmful.

@stephentoub do you have any opinions?

**Comment by stephentoub** [comment]
> @stephentoub do you have any opinions?

As you say, callers shouldn't assume anything about the `ValueTask`s they receive and should code expecting that multiple consumption of such a `ValueTask` is a bug.  This is discussed in detail in https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/.  It's also validated by https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2012.  It'd be fine to incorporate material from that blog post into the `ValueTask` docs, but as multiple folks have stated on this thread, we definitely do not want to update the docs for individual methods to discuss the current implementation of the method and what type is backing the `ValueTask` (just as we don't document for methods strongly-typed to return ISomeInterface exactly what concrete type they currently return that implements that interface).

**Comment by Uzivatel919** [comment]
@stephentoub, I guess I got on same wavelength. I see propagating internals/implementation details is subject to avoidance.

Back to my original swing. I agree with details hiding, still important information like unrecommended repeated calls should be stated in method remarks even if there is [CA2012: Use ValueTasks correctly](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2012). More when in remarks is basically nothing untold. Note also that this analyzer rule could be suppressed or _broken_ (let say by not covered syntax). More resources on this can be always useful.

I see basically that [When to suppress warnings](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2012#when-to-suppress-warnings) explains what I was about to be made explained in Remarks.

So basically I find sufficient to make remarks link either of sources:

- [Understanding the Whys, Whats, and Whens of ValueTask](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/)
- Remarks on struct level pages (https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=net-5.0#remarks, https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask?view=net-5.0#remarks)
- [When to suppress warnings](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2012#when-to-suppress-warnings)

with some brief notice around it.

**Comment by buyaa-n** [comment]
> Back to my original swing. I agree with details hiding, still important information like unrecommended repeated calls should be stated in method remarks even if there is CA2012: Use ValueTasks correctly. More when in remarks is basically nothing untold. Note also that this analyzer rule could be suppressed or broken (let say by not covered syntax). More resources on this can be always useful. 

Sounds like you are suggesting 

1. To add a that `repeated calls are not recommended` into `Remarks` section of `ValueTask.AsTask` and `ValueTask<TResult>.AsTask` section
2. Also add a link to one of the reference docs

Those suggestions sound good to me, could you raise a PR with the suggested changes?

**Comment by stephentoub** [comment]
What specifically needs to be made clearer in the remarks?  They already say this:
```
The following operations should never be performed on a ValueTask instance:
    Awaiting the instance multiple times.
    Calling AsTask multiple times.
    Using more than one of these techniques to consume the instance.
```

**Comment by buyaa-n** [comment]
> What specifically needs to be made clearer in the remarks? They already say this:

No not about those remarks, I thought he is talking about [the API level remarks](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask.astask?view=net-5.0#remarks)
```
Remarks
This method either returns the wrapped task object, if one exists, or it manufactures a new task object to represent the result.
```

**Comment by Uzivatel919** [comment]
Not those on `AsTask` 
 - [`ValueTask.AsTask`|Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask.astask?view=net-5.0#remarks)
- [`ValueTask<TResult>.AsTask`|Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1.astask?view=net-5.0#remarks)

Same applies for `GetAwaiter` while `Property` has some information about this.

**Comment by stephentoub** [comment]
> No not about those remarks, I thought he is talking about the API level remarks

I see.  If we want to propagate some of the info from the type-level remarks to these APIs, that's fine.
